pulsar/transaction_coordinator_client.go

```diff
@@ -20,23 +20,322 @@ package pulsar
-	cons      []internal.Connection
+	handlers  []*transactionHandler
 	epoch     uint64
 	semaphore internal.Semaphore
 	log       log.Logger
 }
 
+type transactionHandler struct {
+	tc              *transactionCoordinatorClient
+	state           uAtomic.Int32
+	conn            uAtomic.Value
+	partition       uint64
+	closeCh         chan any
+	requestCh       chan any
+	connectClosedCh chan *connectionClosed
+	log             log.Logger
+}
+
+type txnHandlerState int
+
+const (
+	txnHandlerReady = iota
+	txnHandlerClosed
+)
+
+func (t *transactionHandler) getState() txnHandlerState {
+	return txnHandlerState(t.state.Load())
+}
+
+func (t *transactionHandler) setState(state txnHandlerState) {
+	t.state.Store(int32(state))
+}
+
+func (tc *transactionCoordinatorClient) newTransactionHandler(partition uint64) (*transactionHandler, error) {
+	handler := &transactionHandler{
+		tc:              tc,
+		partition:       partition,
+		closeCh:         make(chan any),
+		requestCh:       make(chan any),
+		connectClosedCh: make(chan *connectionClosed),
+		log:             tc.log.SubLogger(log.Fields{"txn handler partition": partition}),
+	}
+	err := handler.grabConn()
+	if err != nil {
+		return nil, err
+	}
+	go handler.runEventsLoop()
+	return handler, nil
+}
+
+func (t *transactionHandler) grabConn() error {
+	lr, err := t.tc.client.lookupService.Lookup(getTCAssignTopicName(t.partition))
+	if err != nil {
+		t.log.WithError(err).Warn("Failed to lookup the transaction_impl " +
+			"coordinator assign topic [" + strconv.FormatUint(t.partition, 10) + "]")
+		return err
+	}
+
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	cmdTCConnect := pb.CommandTcClientConnectRequest{
+		RequestId: proto.Uint64(requestID),
+		TcId:      proto.Uint64(t.partition),
+	}
+
+	res, err := t.tc.client.rpcClient.Request(lr.LogicalAddr, lr.PhysicalAddr, requestID,
+		pb.BaseCommand_TC_CLIENT_CONNECT_REQUEST, &cmdTCConnect)
+
+	if err != nil {
+		t.log.WithError(err).Error("Failed to connect transaction_impl coordinator " +
+			strconv.FormatUint(t.partition, 10))
+		return err
+	}
+
+	go func() {
+		select {
+		case <-t.closeCh:
+			return
+		case <-res.Cnx.WaitForClose():
+			t.connectClosedCh <- &connectionClosed{}
+		}
+	}()
+	t.conn.Store(res.Cnx)
+	t.log.Infof("Transaction handler with transaction coordinator id %d connected", t.partition)
+	return nil
+}
+
+func (t *transactionHandler) getConn() internal.Connection {
+	return t.conn.Load().(internal.Connection)
+}
+
+func (t *transactionHandler) runEventsLoop() {
+	for {
+		select {
+		case <-t.closeCh:
+			return
+		case req := <-t.requestCh:
+			switch r := req.(type) {
+			case *newTxnOp:
+				t.newTransaction(r)
+			case *addPublishPartitionOp:
+				t.addPublishPartitionToTxn(r)
+			case *addSubscriptionOp:
+				t.addSubscriptionToTxn(r)
+			case *endTxnOp:
+				t.endTxn(r)
+			}
+		case <-t.connectClosedCh:
+			t.log.Infof("Transaction handler %d will reconnect to the transaction coordinator", t.partition)
```

Comment:
This log seems redundant because you already log "Reconnect to broker" in `reconnectToBroker`. You can add the partition info into that log and remove this line.

```suggestion
```